File: generator_article/22.1.3/iterable.lua
function objectEntries(jane)
    return coroutine.wrap(function()
        for key, value in pairs(jane) do
            coroutine.yield(key, value)
        end
    end)
end

local jane = { first = "Jane", last = "Doe" }
-- local co = objectEntries(jane)
-- print(co())
-- print(co())

-- com for in - simula as chamadas acima
for key, value in objectEntries(jane) do
    print(key, value)
end

File: generator_article/22.1.3/iterable.js
function * objectEntries(obj) {
    const propKeys = Reflect.ownKeys(obj); // Obtém todas as chaves (strings e símbolos) do objeto.
    
    for (const propKey of propKeys) { // Itera sobre as chaves.
        yield [propKey, obj[propKey]]; // "Retorna" um par [chave, valor] por vez.
    }
}

const jane = { first: 'Jane', last: 'Doe', middle: 'Sarah' };
for(const [key, value] of objectEntries(jane)){ // for-of: simula o next()
    console.log(`${key}: ${value}`);

}